include/boost/iostreams/detail/streambuf/indirect_streambuf.hpp

```diff
@@ -421,6 +423,65 @@ void indirect_streambuf<T, Tr, Alloc, Mode>::init_put_area()
         setp(0, 0);
 }
 
+//----------Optimized xsgetn & xsputn-----------------------------------------//
+
+template<typename T, typename Tr, typename Alloc, typename Mode>
+std::streamsize indirect_streambuf<T, Tr, Alloc, Mode>::xsgetn
+    (char_type* s, std::streamsize n)
+{
+    if (n < (std::max)(in().size() - pback_size_ - 2, std::streamsize(2)))
+        return streambuf_type::xsgetn(s, n);
+
+    std::streamsize ret = 0;
+    if (!gptr()) init_get_area();
+    const std::streamsize avail = egptr() - gptr();
+    // First, copy the chars already present in the buffer.
+
+    if (avail != 0) {
+        traits_type::copy(s, gptr(), avail);
+        s += avail;
+        setg(eback(), gptr() + avail, egptr());
+        ret += avail;
+        n -= avail;
+    }
+
+    // Need to loop in case of short reads.
+    for (std::streamsize len; n != 0; n -= len) {
+        len = obj().read(s, n, next_);
+
+        if (len <= 0) {
+            this->set_true_eof(true);
```

Comment:
In IOStreams library the devices return -1 for EOF. Really you can change it to ret == -1. So this is correct.